GitUI/CommandsDialogs/SettingsDialog/ShellExtension/ShellExtensionManager.cs

```diff
@@ -0,0 +1,153 @@
+                    @"*\shellex\ContextMenuHandlers\GitExtensions2", null)) ||
+                string.IsNullOrEmpty(CommonLogic.GetRegistryValue(Registry.ClassesRoot,
+                    @"Directory\shellex\ContextMenuHandlers\GitExtensions2", null)) ||
+                string.IsNullOrEmpty(CommonLogic.GetRegistryValue(Registry.ClassesRoot,
+                    @"Directory\Background\shellex\ContextMenuHandlers\GitExtensions2",
+                    null));
+        }
+
+        /// <summary>
+        /// Register shell extensions
+        /// </summary>
+        /// <exception cref="FileNotFoundException">If at least one necessary for registration file wasn't found</exception>
+        /// <exception cref="Win32Exception">If user canceled elevation dialog</exception>
+        public static void RegisterShellExtension() => RegisterAllShellExtensionDlls("{0}");
+
+        /// <summary>
+        /// Register shell extensions.
+        /// The same as <see cref="RegisterShellExtension"/> but throws no exceptions
+        /// </summary>
+        public static void RegisterShellExtensionSafe()
+        {
+            try
+            {
+                RegisterShellExtension();
+            }
+            catch
+            {
+                // ignored
+            }
+        }
+
+        /// <summary>
+        /// Unregister shell extensions
+        /// </summary>
+        /// <exception cref="FileNotFoundException">If at least one necessary for registration file wasn't found</exception>
+        /// <exception cref="Win32Exception">If user canceled elevation dialog</exception>
+        public static void UnregisterShellExtension() => RegisterAllShellExtensionDlls("/u {0}");
+
+        /// <summary>
+        /// Unregister shell extensions.
+        /// The same as <see cref="UnregisterShellExtension"/> but throws no exceptions
+        /// </summary>
+        public static void UnregisterShellExtensionSafe()
+        {
+            try
+            {
+                UnregisterShellExtension();
+            }
+            catch
+            {
+                // ignored
+            }
+        }
+
+        private static void RegisterAllShellExtensionDlls(string argumentsPattern)
+        {
+            RegisterSingleShellExtensionDll(CommonLogic.GitExtensionsShellEx32Name, argumentsPattern);
+            if (Environment.Is64BitOperatingSystem)
+            {
+                RegisterSingleShellExtensionDll(CommonLogic.GitExtensionsShellEx64Name, argumentsPattern);
+            }
+        }
+
+        private static void RegisterSingleShellExtensionDll(string dllName, string argumentsPattern)
+        {
+            string path = FindFileInBinFolders(dllName);
+            if (string.IsNullOrEmpty(path))
+            {
+                throw new FileNotFoundException(null, dllName);
+            }
+
+            ProcessStartInfo pi = BuildRegisterStartupInfo(string.Format(argumentsPattern, path));
+            Process.Start(pi)?.WaitForExit();
```

Comment:
Will this lock the app while we're waiting for the registration to complete?

Reply:
The code was extracted as is from ChecklistSettingsPage. I decided not to change the logic.
As for the question - indeed, but don't think its an issue because after the call user will see UAC window so waiting here will be OK.